openwisp_radius/tasks.py

```diff
@@ -111,3 +116,89 @@ def send_login_email(accounting_data):
+
+    def get_radsecret_from_radacct(rad_acct):
+        qs = Nas.objects.filter(organization_id=rad_acct.organization_id).only(
+            'name', 'secret'
+        )
+        for nas in qs.iterator():
+            if ipaddress.ip_address(rad_acct.nas_ip_address) in ipaddress.ip_network(
+                nas.name
+            ):
+                return nas.secret
+
+    def get_radius_attributes():
+        attributes = {}
+        rad_group_checks = RadiusGroupCheck.objects.filter(group_id=new_group_id)
+        if rad_group_checks:
+            for check in rad_group_checks:
+                attributes[check.attribute] = f'{check.op}{check.value}'
+        elif (
+            not rad_group_checks
+            and RadiusGroup.objects.filter(id=new_group_id).exists()
+        ):
+            # The new group does not have any limitations.
+            # Unset attributes set by the previous group.
+            rad_group_checks = RadiusGroupCheck.objects.filter(group_id=old_group_id)
+            for check in rad_group_checks:
+                attributes[check.attribute] = ':='
+        return attributes
+
+    try:
+        user = User.objects.get(pk=user_id)
+    except User.DoesNotExist:
+        logger.warning(
+            f'Failed to find user with "{user_id}" ID. Skipping CoA operation.'
+        )
+        return
+    # Check is user has open RadiusAccounting sessions
+    open_sessions = RadiusAccounting.objects.filter(
+        username=user.username, stop_time__isnull=True
+    ).select_related('organization', 'organization__radius_settings')
+    if not open_sessions:
+        logger.warning(
+            f'The user with "{user_id}" ID does not have any open'
+            ' RadiusAccounting sessions. Skipping CoA operation.'
+        )
+        return
+
+    attributes = get_radius_attributes()
+    if not attributes:
+        logger.warning(
+            f'Failed to find RadiusGroup with "{new_group_id}".'
+            ' Skipping CoA operation.'
+        )
+        return
+
+    attributes['User-Name'] = user.username
+    for session in open_sessions:
+        if not session.organization.radius_settings.get_setting('coa_enabled'):
+            continue
+        radsecret = get_radsecret_from_radacct(session)
+        if not radsecret:
+            logger.warning(
+                f'Failed to find RADIUS secret for "{session.unique_id}"'
+                ' RadiusAccounting object. Skipping CoA operation'
+                ' for this session.'
+            )
+            continue
+        client = RadClient(
+            host=session.nas_ip_address,
+            radsecret=radsecret,
+            dicts=app_settings.RADCLIENT_ATTRIBUTE_DICTIONARIES,
+        )
+        result = client.perform_change_of_authorization(attributes)
+        if not result:
+            logger.warning(
+                f'Failed to perform CoA for "{session.unique_id}"'
+                f' RadiusAccounting object of "{user}" user'
+            )
```

Comment:
I see the logging here but I think it's better to do it above as I recommended so we know whether the failure is due to timeout or a coa nack,